src/buckets/components/createBucketForm/MeasurementSchemaUtils.ts

```diff
@@ -0,0 +1,75 @@
+import {trim} from 'lodash'
+
+const typeStrings = ['timestamp', 'tag', 'field']
+const dataTypeStrings = ['integer', 'float', 'boolean', 'string', 'unsigned']
+
+/**
+ * manually checking that the literal string union types are correct
+ *
+ * our types are generated, and changing it so there is an exported array that is *then* used
+ * for literal union type creation is out of scope.
+ * https://newbedev.com/checking-validity-of-string-literal-union-type-at-runtime
+ *
+ * right now, just checking true/false; is it valid?  later on, may add a message saying which part(s) are invalid
+ *
+ * The backend checks for uniqueness among the schema names and among the column names within each schema.
+ * it gives us a relevant error message if the uniqueness fails.
+ *
+ * The backend also checks if the columns themselves are proper, there must be at least two columns:
+ *   1) a time field of time 'timestamp'
+ *   2) a field with type 'field'
+ *
+ *   not checking for those at this time on the frontend, letting the backend do the checking;
+ *   the code propagates up the error messages to show the user.
+ * */
+export const areColumnsKosher = columns => {
+  if (Array.isArray(columns)) {
+    const validArray = columns.map(item => {
+      // for each item, does it have the necessary stuff?
+      const hasName = 'name' in item && typeof item.name === 'string'
+      const hasType = 'type' in item && typeStrings.includes(item.type)
+
+      // the semi-optional part: if it isn't there; it is fine; but if it is check it!
+      // BUT, if the type is 'field' it is required!
+      let dataTypePartIsValid = true
+
+      if ('dataType' in item) {
+        dataTypePartIsValid = dataTypeStrings.includes(item.dataType)
+      } else {
+        // not there:
+        if (item.type === 'field') {
+          dataTypePartIsValid = false
+        }
+      }
+      return hasName && hasType && dataTypePartIsValid
+    })
+
+    return validArray.reduce((prevVal, curVal) => prevVal && curVal, true)
+  }
```

Comment:
I'm trying to follow this functional logic and it seems like it can be simplified and made more efficient (currently it's n * n) by not using `map` and `reduce`. If this was a basic `for` loop, any time you encountered a bad value, you could just `return false` from the entire loop and short circuit needing to finish looping over the rest of the array and then looping over it again. 

```ts
if (Array.isArray(columns)) {
  for (let i = 0; i < columns.length; i++) {
    if (!('name' in item && typeof item.name === 'string') {
      return false
    }

    if (!('type' in item && typeStrings.includes(item.type))) {
      return false
    }

    if ('dataType' in item) {
      if (!dataTypeStrings.includes(item.dataType)) {
        return false
      }
    } else if (item.type === 'field') {
      return false
    }
  }
  return true  
}

return false
```

is this not equivalent in logic but simpler in terms of execution?

Reply:
unfortunately that will not work.  if you return something from within a for loop, it just keeps looping.  but using the every() method, similar to the some() method, is the way to go;  for example; this code which I just ran in my console:

```
let bar = [true, true, false, true, true, false]

let checkIt = (arr) => {
for(let i=0; i<arr.length; i++){
    if (!arr){
        console.log(`at index ${i}`)
        return false;
    }
}
return true;
}

checkIt(bar)

let checkIt2= (arr) => {
    let allGood = (x) => x;
    return arr.every(allGood)
}

console.log('checkit:', checkIt(bar))
console.log("checkit2:", checkIt2(bar))
```

the first console.log returns true, the second returns false (false is the correct answer here).
so I'll switch it to 'every'

Reply:
keeping the variable names (hasName, hasType) for clarity

Reply:
>however, the 'every' function (similar to the some function) will work too. so will use that.

cool, works for me. Also I realized after I wrote this review that this is a `n + n` not `n*n` loop. Minor detail but I wanted to correct it.